File: artifacts/api-server/src/lib/positionSizing.ts
import type { BtcRegime, PositionSide, TradeOutcome } from "./adaptiveEngine";
import type { BotConfig } from "./botConfig";
import type { SymbolRotationRank, SymbolRotationState } from "./symbolRotation";

export type PositionRiskTier = "MICRO" | "SCOUT" | "BASE" | "BOOST" | "AGGRESSIVE" | "MAX_SNIPER";

export interface PositionSizingConfig {
  enabled: boolean;
  minMargin: number;
  baseRiskPct: number;
  maxRiskPctPerTrade: number;
  maxTotalRiskPct: number;
  maxSymbolRiskPct: number;
  maxConcurrentPositions: number;
  maxMarginUtilization: number;
  demoLearningAggressive: boolean;
}

export interface PositionSizingOpenPosition {
  symbol: string;
  positionSide?: PositionSide;
  marginUsed: number;
  leverage: number;
}

export interface PositionSizingInput {
  symbol: string;
  positionSide: PositionSide;
  accountEquity: number;
  availableMargin: number;
  aggressiveScore: number;
  calibratedScore?: number;
  coachRank?: number;
  rotationState?: SymbolRotationState;
  aggressionState?: "DEFENSIVE" | "NORMAL" | "AGGRESSIVE" | "SNIPER";
  recentPnl?: number;
  recentWinRate?: number;
  profitFactor?: number;
  drawdown?: number;
  executionSlippageBps?: number;
  campaignDepth?: number;
  campaignPnl?: number;
  previousEntryMargin?: number;
  currentOpenPositions?: PositionSizingOpenPosition[];
  symbolConcentration?: number;
  sideContextWinRate?: number;
  sideContextProfitFactor?: number;
  experimentArm?: string;
  dataQualityDegraded?: boolean;
  exitPreservingProfit?: boolean;
  baseMarginFallback?: number;
  leverageFallback?: number;
  stopLossPct?: number;
  config?: Partial<PositionSizingConfig>;
}

export interface PositionSizingDecision {
  recommendedMargin: number;
  recommendedLeverage: number;
  riskTier: PositionRiskTier | "NO_TRADE";
  sizeMultiplier: number;
  reason: string;
  maxLossIfStop: number;
  notional: number;
  baseMargin: number;
  approved: boolean;
  gateRejects: string[];
  diagnostics: {
    totalRiskPct: number;
    symbolRiskPct: number;
    marginUtilizationAfter: number;
    recentWinRate: number;
    profitFactor: number;
    drawdownPct: number;
    executionSlippageBps: number;
  };
}

const TIER_MULTIPLIER: Record<PositionRiskTier, number> = {
  MICRO: 0.25,
  SCOUT: 0.5,
  BASE: 1.0,
  BOOST: 1.25,
  AGGRESSIVE: 1.5,
  MAX_SNIPER: 2.0,
};

function envNum(key: string, fallback: number): number {
  const value = process.env[key];
  if (!value) return fallback;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function envBool(key: string, fallback: boolean): boolean {
  const value = process.env[key];
  if (!value) return fallback;
  return value.toLowerCase() === "true" || value === "1";
}

function clamp(value: number, min: number, max: number): number {
  if (!Number.isFinite(value)) return min;
  return Math.max(min, Math.min(max, value));
}

function round(value: number, digits = 4): number {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

function safePct(value: number | undefined): number {
  if (!Number.isFinite(value ?? NaN)) return 0;
  return Math.max(0, value!);
}

export function getPositionSizingConfig(botConfig?: BotConfig): PositionSizingConfig {
  const aggressiveMode = botConfig?.activeMode === "aggressive";
  const demoLearning = String(process.env["EXPERIMENT_ARM"] ?? "").toLowerCase().includes("demo_learning_aggressive");
  return {
    enabled: envBool("POSITION_SIZING_ENABLED", true),
    minMargin: envNum("MIN_MARGIN", Math.max(0.1, botConfig?.marginPerTrade ?? 5)),
    baseRiskPct: envNum("BASE_RISK_PCT", 0.005),
    maxRiskPctPerTrade: envNum("MAX_RISK_PCT_PER_TRADE", aggressiveMode || demoLearning ? 0.025 : 0.015),
    maxTotalRiskPct: envNum("MAX_TOTAL_RISK_PCT", aggressiveMode || demoLearning ? 0.12 : 0.08),
    maxSymbolRiskPct: envNum("MAX_SYMBOL_RISK_PCT", aggressiveMode || demoLearning ? 0.05 : 0.03),
    maxConcurrentPositions: botConfig?.maxConcurrentPositions ?? envNum("SCALP_MAX_CONCURRENT_POSITIONS", 10),
    maxMarginUtilization: botConfig?.maxMarginUtilization ?? envNum("SCALP_MAX_MARGIN_UTILIZATION", 0.5),
    demoLearningAggressive: aggressiveMode || demoLearning,
  };
}

function initialTier(score: number, rotationState?: SymbolRotationState, slippageBps = 0): PositionRiskTier | "NO_TRADE" {
  if (score < 0.55) return "NO_TRADE";
  if (score >= 0.85 && rotationState === "HOT" && slippageBps <= 3) return "MAX_SNIPER";
  if (score >= 0.78) return rotationState === "REDUCED" ? "BOOST" : "AGGRESSIVE";
  if (score >= 0.68) return rotationState === "REDUCED" ? "BASE" : "BOOST";
  return score >= 0.62 ? "BASE" : "SCOUT";
}

function shiftTier(tier: PositionRiskTier, delta: number): PositionRiskTier {
  const tiers: PositionRiskTier[] = ["MICRO", "SCOUT", "BASE", "BOOST", "AGGRESSIVE", "MAX_SNIPER"];
  return tiers[clamp(tiers.indexOf(tier) + delta, 0, tiers.length - 1)];
}

export function calculatePositionSizing(input: PositionSizingInput): PositionSizingDecision {
  const cfg = { ...getPositionSizingConfig(), ...(input.config ?? {}) };
  const equity = Math.max(0, input.accountEquity || 0);
  const leverage = Math.max(1, Math.round(input.leverageFallback ?? 14));
  const stopLossPct = Math.max(0.01, input.stopLossPct ?? 0.1);
  const baseMargin = Math.max(cfg.minMargin, equity * cfg.baseRiskPct, input.baseMarginFallback ?? 0);
  const slippageBps = safePct(input.executionSlippageBps);
  const openPositions = input.currentOpenPositions ?? [];
  const gateRejects: string[] = [];
  const reasons: string[] = [];
  const sizingScore = clamp(input.calibratedScore ?? input.aggressiveScore, 0, 1);

  if (!cfg.enabled) {
    const margin = Math.max(cfg.minMargin, input.baseMarginFallback ?? baseMargin);
    return buildDecision("BASE", 1, margin, leverage, stopLossPct, baseMargin, true, ["fixed_sizing_disabled"], [], equity, openPositions, input);
  }

  let tier = initialTier(sizingScore, input.rotationState, slippageBps);
  if (tier === "NO_TRADE") {
    return buildDecision("NO_TRADE", 0, 0, leverage, stopLossPct, baseMargin, false, ["score_below_trade_floor"], ["SCORE_REJECT: calibratedScore < 0.55"], equity, openPositions, input);
  }
  if (input.calibratedScore !== undefined && Math.abs(input.calibratedScore - input.aggressiveScore) >= 0.03) {
    reasons.push("score_calibrated");
  }

  if (input.rotationState === "HOT") reasons.push("hot_symbol");
  if (input.rotationState === "REDUCED" || input.rotationState === "RECOVERY") {
    tier = shiftTier(tier, -2);
    reasons.push("symbol_reduced");
  }
  if (input.aggressionState === "DEFENSIVE") {
    tier = shiftTier(tier, -2);
    reasons.push("defensive_state");
  }
  if ((input.profitFactor ?? 1) >= 1.6 && (input.recentWinRate ?? 0.5) >= 0.56) {
    tier = shiftTier(tier, 1);
    reasons.push("recent_edge_good");
  }
  if (input.exitPreservingProfit) {
    tier = shiftTier(tier, 1);
    reasons.push("exit_preserving_profit");
  }
  if ((input.drawdown ?? 0) > equity * 0.02 || (input.drawdown ?? 0) > 2) {
    tier = shiftTier(tier, -1);
    reasons.push("drawdown_elevated");
  }
  if (slippageBps > 6) {
    tier = shiftTier(tier, -1);
    reasons.push("slippage_high");
  }
  if (input.dataQualityDegraded) {
    tier = shiftTier(tier, -2);
    reasons.push("data_quality_degraded");
  }
  if ((input.sideContextWinRate ?? 0.5) < 0.45 && (input.sideContextProfitFactor ?? 1) < 1) {
    tier = shiftTier(tier, -1);
    reasons.push("side_context_weak");
  }

  const depth = Math.max(1, Math.floor(input.campaignDepth ?? 1));
  if (depth === 2) {
    tier = shiftTier(tier, -1);
    reasons.push("depth_2_no_pyramid");
  } else if (depth >= 3 && (input.campaignPnl ?? 0) <= 0) {
    tier = shiftTier(tier, -2);
    reasons.push("deep_negative_campaign");
  }

  let multiplier = TIER_MULTIPLIER[tier];
  let margin = baseMargin * multiplier;
  const previous = input.previousEntryMargin;
  if (depth > 1 && previous !== undefined && (input.campaignPnl ?? 0) <= 0 && margin > previous) {
    margin = previous;
    multiplier = margin / Math.max(baseMargin, 0.000001);
    tier = shiftTier(tier, -1);
    reasons.push("martingale_guard");
  }
  if (depth > 1 && previous !== undefined && (input.campaignPnl ?? 0) > 0 && sizingScore < 0.78 && margin > previous) {
    margin = previous;
    multiplier = margin / Math.max(baseMargin, 0.000001);
    reasons.push("stack_score_not_higher");
  }

  const maxPerTradeMargin = equity > 0
    ? (equity * cfg.maxRiskPctPerTrade) / (leverage * (stopLossPct / 100))
    : margin;
  if (margin > maxPerTradeMargin) {
    margin = maxPerTradeMargin;
    multiplier = margin / Math.max(baseMargin, 0.000001);
    reasons.push("per_trade_risk_cap");
  }

  if (margin > input.availableMargin) {
    margin = Math.max(0, input.availableMargin);
    multiplier = margin / Math.max(baseMargin, 0.000001);
    reasons.push("available_margin_cap");
  }

  const risk = margin * leverage * (stopLossPct / 100);
  const openRisk = openPositions.reduce((sum, pos) => sum + pos.marginUsed * pos.leverage * (stopLossPct / 100), 0);
  const symbolRisk = openPositions
    .filter((pos) => pos.symbol.toUpperCase() === input.symbol.toUpperCase())
    .reduce((sum, pos) => sum + pos.marginUsed * pos.leverage * (stopLossPct / 100), 0) + risk;
  const usedMargin = openPositions.reduce((sum, pos) => sum + pos.marginUsed, 0);

  const totalRiskPct = equity > 0 ? (openRisk + risk) / equity : 0;
  const symbolRiskPct = equity > 0 ? symbolRisk / equity : 0;
  const marginUtilizationAfter = equity > 0 ? (usedMargin + margin) / equity : 0;

  if (openPositions.length >= cfg.maxConcurrentPositions) gateRejects.push("SIZE_CAPITAL_REJECT: max concurrent positions");
  if (totalRiskPct > cfg.maxTotalRiskPct) gateRejects.push("SIZE_TOTAL_RISK_REJECT");
  if (symbolRiskPct > cfg.maxSymbolRiskPct) gateRejects.push("SIZE_SYMBOL_RISK_REJECT");
  if (marginUtilizationAfter > cfg.maxMarginUtilization) gateRejects.push("SIZE_MARGIN_UTILIZATION_REJECT");
  if (margin <= 0) gateRejects.push("SIZE_MARGIN_REJECT");

  if (reasons.length === 0) reasons.push(`${tier.toLowerCase()}_score_band`);
  return buildDecision(tier, multiplier, margin, leverage, stopLossPct, baseMargin, gateRejects.length === 0, reasons, gateRejects, equity, openPositions, input);
}

function buildDecision(
  tier: PositionRiskTier | "NO_TRADE",
  multiplier: number,
  margin: number,
  leverage: number,
  stopLossPct: number,
  baseMargin: number,
  approved: boolean,
  reasons: string[],
  gateRejects: string[],
  equity: number,
  openPositions: PositionSizingOpenPosition[],
  input: PositionSizingInput,
): PositionSizingDecision {
  const maxLossIfStop = margin * leverage * (stopLossPct / 100);
  const symbol = input.symbol.toUpperCase();
  const openRisk = openPositions.reduce((sum, pos) => sum + pos.marginUsed * pos.leverage * (stopLossPct / 100), 0);
  const symbolRisk = openPositions
    .filter((pos) => pos.symbol.toUpperCase() === symbol)
    .reduce((sum, pos) => sum + pos.marginUsed * pos.leverage * (stopLossPct / 100), 0) + maxLossIfStop;
  const usedMargin = openPositions.reduce((sum, pos) => sum + pos.marginUsed, 0);
  return {
    recommendedMargin: round(Math.max(0, margin), 4),
    recommendedLeverage: leverage,
    riskTier: tier,
    sizeMultiplier: round(Math.max(0, multiplier), 4),
    reason: reasons.join("_"),
    maxLossIfStop: round(maxLossIfStop, 4),
    notional: round(margin * leverage, 4),
    baseMargin: round(baseMargin, 4),
    approved,
    gateRejects,
    diagnostics: {
      totalRiskPct: round(equity > 0 ? (openRisk + maxLossIfStop) / equity : 0, 6),
      symbolRiskPct: round(equity > 0 ? symbolRisk / equity : 0, 6),
      marginUtilizationAfter: round(equity > 0 ? (usedMargin + margin) / equity : 0, 6),
      recentWinRate: round(input.recentWinRate ?? 0.5, 4),
      profitFactor: round(input.profitFactor ?? 1, 4),
      drawdownPct: round(equity > 0 ? (input.drawdown ?? 0) / equity : 0, 6),
      executionSlippageBps: round(input.executionSlippageBps ?? 0, 4),
    },
  };
}

export function summarizeSizingStatus(input: {
  equity: number;
  outcomes: TradeOutcome[];
  openPositions: PositionSizingOpenPosition[];
  config: BotConfig;
  btcRegime?: BtcRegime;
}): Record<string, unknown> {
  const cfg = getPositionSizingConfig(input.config);
  const baseMargin = Math.max(cfg.minMargin, input.equity * cfg.baseRiskPct, input.config.marginPerTrade);
  const byTier = new Map<string, { trades: number; wins: number; pnl: number; maxDrawdown: number; equityCurve: number[] }>();
  for (const outcome of input.outcomes) {
    const tier = outcome.riskTier ?? "UNSIZED";
    const row = byTier.get(tier) ?? { trades: 0, wins: 0, pnl: 0, maxDrawdown: 0, equityCurve: [] };
    row.trades += 1;
    row.wins += outcome.realizedPnl > 0 ? 1 : 0;
    row.pnl += outcome.realizedPnl;
    row.equityCurve.push(row.pnl);
    let peak = 0;
    row.maxDrawdown = Math.max(...row.equityCurve.map((value) => {
      peak = Math.max(peak, value);
      return peak - value;
    }), 0);
    byTier.set(tier, row);
  }
  const tierStats = Object.fromEntries([...byTier.entries()].map(([tier, row]) => [tier, {
    trades: row.trades,
    pnl: round(row.pnl, 4),
    winRate: row.trades > 0 ? round(row.wins / row.trades, 4) : 0,
    drawdown: round(row.maxDrawdown, 4),
  }]));
  const concentration = new Map<string, number>();
  for (const pos of input.openPositions) {
    concentration.set(pos.symbol.toUpperCase(), (concentration.get(pos.symbol.toUpperCase()) ?? 0) + pos.marginUsed * pos.leverage);
  }
  const largest = [...concentration.entries()].sort((a, b) => b[1] - a[1])[0] ?? null;
  return {
    enabled: cfg.enabled,
    baseMargin: round(baseMargin, 4),
    equityUsed: round(input.equity, 4),
    sizeTierDistribution: Object.fromEntries([...byTier.entries()].map(([tier, row]) => [tier, row.trades])),
    pnlByRiskTier: Object.fromEntries([...byTier.entries()].map(([tier, row]) => [tier, round(row.pnl, 4)])),
    winRateByRiskTier: Object.fromEntries([...byTier.entries()].map(([tier, row]) => [tier, row.trades > 0 ? round(row.wins / row.trades, 4) : 0])),
    drawdownByRiskTier: Object.fromEntries([...byTier.entries()].map(([tier, row]) => [tier, round(row.maxDrawdown, 4)])),
    tierStats,
    largestSymbolConcentration: largest ? { symbol: largest[0], notional: round(largest[1], 4) } : null,
    compoundingCurve: input.outcomes
      .slice(-200)
      .reduce<Array<{ trade: number; equity: number; baseMargin: number }>>((rows, outcome, index) => {
        const prev = rows[index - 1]?.equity ?? input.equity;
        const equity = prev + outcome.realizedPnl;
        rows.push({ trade: index + 1, equity: round(equity, 4), baseMargin: round(Math.max(cfg.minMargin, equity * cfg.baseRiskPct), 4) });
        return rows;
      }, []),
    recommendedConfig: cfg,
  };
}

export function recentSizingStats(outcomes: TradeOutcome[], symbol: string, side: PositionSide, now = Date.now()): {
  recentPnl: number;
  recentWinRate: number;
  profitFactor: number;
  drawdown: number;
  executionSlippageBps: number;
  sideContextWinRate: number;
  sideContextProfitFactor: number;
} {
  const since = now - 4 * 60 * 60_000;
  const recent = outcomes.filter((outcome) => (outcome.exitTime ?? outcome.entryTime) >= since);
  const scoped = recent.filter((outcome) => outcome.symbol.toUpperCase() === symbol.toUpperCase() && outcome.positionSide === side);
  const calc = scoped.length > 0 ? scoped : recent;
  const wins = calc.filter((outcome) => outcome.realizedPnl > 0);
  const losses = calc.filter((outcome) => outcome.realizedPnl <= 0);
  const grossWin = wins.reduce((sum, outcome) => sum + outcome.realizedPnl, 0);
  const grossLoss = Math.abs(losses.reduce((sum, outcome) => sum + outcome.realizedPnl, 0));
  let cumulative = 0;
  let peak = 0;
  let drawdown = 0;
  for (const outcome of [...calc].sort((a, b) => (a.exitTime ?? a.entryTime) - (b.exitTime ?? b.entryTime))) {
    cumulative += outcome.realizedPnl;
    peak = Math.max(peak, cumulative);
    drawdown = Math.max(drawdown, peak - cumulative);
  }
  return {
    recentPnl: round(calc.reduce((sum, outcome) => sum + outcome.realizedPnl, 0), 4),
    recentWinRate: calc.length > 0 ? wins.length / calc.length : 0.5,
    profitFactor: grossLoss > 0 ? grossWin / grossLoss : grossWin > 0 ? 999 : 1,
    drawdown,
    executionSlippageBps: calc.length > 0
      ? calc.reduce((sum, outcome) => sum + (outcome.slippagePctNotional ?? 0) * 10_000, 0) / calc.length
      : 0,
    sideContextWinRate: scoped.length > 0 ? scoped.filter((outcome) => outcome.realizedPnl > 0).length / scoped.length : 0.5,
    sideContextProfitFactor: scoped.length > 0
      ? (() => {
          const sw = scoped.reduce((sum, outcome) => sum + Math.max(0, outcome.realizedPnl), 0);
          const sl = Math.abs(scoped.reduce((sum, outcome) => sum + Math.min(0, outcome.realizedPnl), 0));
          return sl > 0 ? sw / sl : sw > 0 ? 999 : 1;
        })()
      : 1,
  };
}

File: artifacts/quant-brain/core/position_sizing.py
"""
Capital compounding / position sizing engine.

The engine scales margin from realized account equity and current edge quality.
It never increases size because the previous trade lost.
"""
from __future__ import annotations

import os
import time
from dataclasses import dataclass, asdict
from typing import Any, Literal

RiskTier = Literal["MICRO", "SCOUT", "BASE", "BOOST", "AGGRESSIVE", "MAX_SNIPER"]

TIER_MULTIPLIER: dict[RiskTier, float] = {
    "MICRO": 0.25,
    "SCOUT": 0.50,
    "BASE": 1.00,
    "BOOST": 1.25,
    "AGGRESSIVE": 1.50,
    "MAX_SNIPER": 2.00,
}
TIER_ORDER: list[RiskTier] = ["MICRO", "SCOUT", "BASE", "BOOST", "AGGRESSIVE", "MAX_SNIPER"]


def _num(value: Any, fallback: float = 0.0) -> float:
    try:
        return float(value) if value is not None else fallback
    except Exception:
        return fallback


def _bool_env(key: str, fallback: bool) -> bool:
    raw = os.environ.get(key)
    if raw is None:
        return fallback
    return raw.lower() in ("1", "true", "yes", "on")


def _env_num(key: str, fallback: float) -> float:
    return _num(os.environ.get(key), fallback)


def _clamp(value: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, value))


def _round(value: float, digits: int = 4) -> float:
    return round(float(value), digits)


@dataclass(frozen=True)
class PositionSizingConfig:
    enabled: bool
    min_margin: float
    base_risk_pct: float
    max_risk_pct_per_trade: float
    max_total_risk_pct: float
    max_symbol_risk_pct: float
    max_concurrent_positions: int
    max_margin_utilization: float
    demo_learning_aggressive: bool


def load_config(config: dict | None = None) -> PositionSizingConfig:
    config = config or {}
    arm = str(config.get("experimentArm") or os.environ.get("EXPERIMENT_ARM", "")).lower()
    aggressive = arm == "demo_learning_aggressive" or str(config.get("activeMode", "")).lower() == "aggressive"
    return PositionSizingConfig(
        enabled=_bool_env("POSITION_SIZING_ENABLED", True),
        min_margin=_env_num("MIN_MARGIN", _num(config.get("marginPerTrade"), 5.0)),
        base_risk_pct=_env_num("BASE_RISK_PCT", 0.005),
        max_risk_pct_per_trade=_env_num("MAX_RISK_PCT_PER_TRADE", 0.025 if aggressive else 0.015),
        max_total_risk_pct=_env_num("MAX_TOTAL_RISK_PCT", 0.12 if aggressive else 0.08),
        max_symbol_risk_pct=_env_num("MAX_SYMBOL_RISK_PCT", 0.05 if aggressive else 0.03),
        max_concurrent_positions=int(_env_num("MAX_CONCURRENT_POSITIONS", _num(config.get("maxConcurrentPositions"), 10))),
        max_margin_utilization=_env_num("MAX_MARGIN_UTILIZATION", _num(config.get("maxMarginUtilization"), 0.5)),
        demo_learning_aggressive=aggressive,
    )


def _initial_tier(score: float, rotation_state: str, slippage_bps: float) -> RiskTier | str:
    if score < 0.55:
        return "NO_TRADE"
    if score >= 0.85 and rotation_state == "HOT" and slippage_bps <= 3:
        return "MAX_SNIPER"
    if score >= 0.78:
        return "BOOST" if rotation_state == "REDUCED" else "AGGRESSIVE"
    if score >= 0.68:
        return "BASE" if rotation_state == "REDUCED" else "BOOST"
    return "BASE" if score >= 0.62 else "SCOUT"


def _shift(tier: RiskTier, delta: int) -> RiskTier:
    idx = TIER_ORDER.index(tier)
    return TIER_ORDER[int(_clamp(idx + delta, 0, len(TIER_ORDER) - 1))]


def calculate_position_size(payload: dict) -> dict:
    cfg = load_config(payload.get("config") or {})
    equity = max(0.0, _num(payload.get("accountEquity") or payload.get("equity"), 0.0))
    available = _num(payload.get("availableMargin"), float("inf"))
    leverage = max(1, int(round(_num(payload.get("leverageFallback") or payload.get("leverage"), 14))))
    stop_loss_pct = max(0.01, _num(payload.get("stopLossPct"), 0.10))
    score = _clamp(_num(payload.get("aggressiveScore"), 0.0), 0.0, 1.0)
    base_margin = max(cfg.min_margin, equity * cfg.base_risk_pct, _num(payload.get("baseMarginFallback"), 0.0))
    slippage_bps = max(0.0, _num(payload.get("executionSlippageBps"), 0.0))
    reasons: list[str] = []
    rejects: list[str] = []

    if not cfg.enabled:
        return _decision("BASE", 1.0, base_margin, leverage, stop_loss_pct, base_margin, True, ["fixed_sizing_disabled"], [], payload)

    tier = _initial_tier(score, str(payload.get("symbolRotationState") or payload.get("rotationState") or "ACTIVE").upper(), slippage_bps)
    if tier == "NO_TRADE":
        return _decision("NO_TRADE", 0.0, 0.0, leverage, stop_loss_pct, base_margin, False, ["score_below_trade_floor"], ["SCORE_REJECT"], payload)

    rotation = str(payload.get("symbolRotationState") or payload.get("rotationState") or "ACTIVE").upper()
    if rotation == "HOT":
        reasons.append("hot_symbol")
    if rotation in ("REDUCED", "RECOVERY"):
        tier = _shift(tier, -2)
        reasons.append("symbol_reduced")
    if str(payload.get("aggressionState") or "").upper() == "DEFENSIVE":
        tier = _shift(tier, -2)
        reasons.append("defensive_state")
    if _num(payload.get("profitFactor"), 1.0) >= 1.6 and _num(payload.get("recentWinRate"), 0.5) >= 0.56:
        tier = _shift(tier, 1)
        reasons.append("recent_edge_good")
    if payload.get("exitPreservingProfit"):
        tier = _shift(tier, 1)
        reasons.append("exit_preserving_profit")
    if _num(payload.get("drawdown"), 0.0) > max(2.0, equity * 0.02):
        tier = _shift(tier, -1)
        reasons.append("drawdown_elevated")
    if slippage_bps > 6:
        tier = _shift(tier, -1)
        reasons.append("slippage_high")
    if payload.get("dataQualityDegraded"):
        tier = _shift(tier, -2)
        reasons.append("data_quality_degraded")

    depth = max(1, int(_num(payload.get("campaignDepth"), 1)))
    if depth == 2:
        tier = _shift(tier, -1)
        reasons.append("depth_2_no_pyramid")
    elif depth >= 3 and _num(payload.get("campaignPnl"), 0.0) <= 0:
        tier = _shift(tier, -2)
        reasons.append("deep_negative_campaign")

    multiplier = TIER_MULTIPLIER[tier]
    margin = base_margin * multiplier
    previous = payload.get("previousEntryMargin")
    if depth > 1 and previous is not None and _num(payload.get("campaignPnl"), 0.0) <= 0 and margin > _num(previous):
        margin = _num(previous)
        multiplier = margin / max(base_margin, 0.000001)
        reasons.append("martingale_guard")

    max_margin_by_trade_risk = (equity * cfg.max_risk_pct_per_trade) / (leverage * (stop_loss_pct / 100)) if equity > 0 else margin
    if margin > max_margin_by_trade_risk:
        margin = max_margin_by_trade_risk
        multiplier = margin / max(base_margin, 0.000001)
        reasons.append("per_trade_risk_cap")
    if margin > available:
        margin = max(0.0, available)
        multiplier = margin / max(base_margin, 0.000001)
        reasons.append("available_margin_cap")

    open_positions = payload.get("currentOpenPositions") or []
    risk = margin * leverage * (stop_loss_pct / 100)
    open_risk = sum(_num(p.get("marginUsed")) * _num(p.get("leverage"), leverage) * (stop_loss_pct / 100) for p in open_positions if isinstance(p, dict))
    symbol = str(payload.get("symbol", "")).upper()
    symbol_risk = risk + sum(
        _num(p.get("marginUsed")) * _num(p.get("leverage"), leverage) * (stop_loss_pct / 100)
        for p in open_positions
        if isinstance(p, dict) and str(p.get("symbol", "")).upper() == symbol
    )
    used_margin = sum(_num(p.get("marginUsed")) for p in open_positions if isinstance(p, dict))
    total_risk_pct = (open_risk + risk) / equity if equity > 0 else 0
    symbol_risk_pct = symbol_risk / equity if equity > 0 else 0
    margin_utilization_after = (used_margin + margin) / equity if equity > 0 else 0
    if len(open_positions) >= cfg.max_concurrent_positions:
        rejects.append("SIZE_CAPITAL_REJECT")
    if total_risk_pct > cfg.max_total_risk_pct:
        rejects.append("SIZE_TOTAL_RISK_REJECT")
    if symbol_risk_pct > cfg.max_symbol_risk_pct:
        rejects.append("SIZE_SYMBOL_RISK_REJECT")
    if margin_utilization_after > cfg.max_margin_utilization:
        rejects.append("SIZE_MARGIN_UTILIZATION_REJECT")
    if margin <= 0:
        rejects.append("SIZE_MARGIN_REJECT")

    if not reasons:
        reasons.append(f"{str(tier).lower()}_score_band")
    return _decision(tier, multiplier, margin, leverage, stop_loss_pct, base_margin, not rejects, reasons, rejects, payload)


def _decision(tier: str, multiplier: float, margin: float, leverage: int, stop_loss_pct: float, base_margin: float,
              approved: bool, reasons: list[str], rejects: list[str], payload: dict) -> dict:
    max_loss = margin * leverage * (stop_loss_pct / 100)
    return {
        "recommendedMargin": _round(max(0.0, margin)),
        "recommendedLeverage": leverage,
        "riskTier": tier,
        "sizeMultiplier": _round(max(0.0, multiplier)),
        "reason": "_".join(reasons),
        "sizeReason": "_".join(reasons),
        "maxLossIfStop": _round(max_loss),
        "notional": _round(margin * leverage),
        "baseMargin": _round(base_margin),
        "approved": approved,
        "gateRejects": rejects,
        "diagnostics": {
            "score": _round(_num(payload.get("aggressiveScore")), 4),
            "recentWinRate": _round(_num(payload.get("recentWinRate"), 0.5), 4),
            "profitFactor": _round(_num(payload.get("profitFactor"), 1.0), 4),
            "generatedAt": time.time(),
        },
    }


def build_status(trades: list[dict], equity: float, config: dict | None = None) -> dict:
    cfg = load_config(config or {})
    base_margin = max(cfg.min_margin, equity * cfg.base_risk_pct, _num((config or {}).get("marginPerTrade"), 0.0))
    by_tier: dict[str, dict[str, float]] = {}
    curve = []
    running_equity = equity
    for idx, trade in enumerate(trades[-500:], start=1):
        tier = str(trade.get("riskTier") or "UNSIZED")
        row = by_tier.setdefault(tier, {"trades": 0, "wins": 0, "pnl": 0.0, "drawdown": 0.0, "peak": 0.0, "cum": 0.0})
        pnl = _num(trade.get("realizedPnl") or trade.get("pnl_usdt"))
        row["trades"] += 1
        row["wins"] += 1 if pnl > 0 else 0
        row["pnl"] += pnl
        row["cum"] += pnl
        row["peak"] = max(row["peak"], row["cum"])
        row["drawdown"] = max(row["drawdown"], row["peak"] - row["cum"])
        running_equity += pnl
        curve.append({"trade": idx, "equity": _round(running_equity), "baseMargin": _round(max(cfg.min_margin, running_equity * cfg.base_risk_pct))})
    return {
        "enabled": cfg.enabled,
        "baseMargin": _round(base_margin),
        "equityUsed": _round(equity),
        "sizeTierDistribution": {k: int(v["trades"]) for k, v in by_tier.items()},
        "pnlByRiskTier": {k: _round(v["pnl"]) for k, v in by_tier.items()},
        "winRateByRiskTier": {k: _round(v["wins"] / v["trades"], 4) if v["trades"] else 0 for k, v in by_tier.items()},
        "drawdownByRiskTier": {k: _round(v["drawdown"]) for k, v in by_tier.items()},
        "compoundingCurve": curve,
        "recommendedConfig": asdict(cfg),
    }
